_posts/wasmer-go-embedding-1.0.md

```diff
@@ -0,0 +1,534 @@
+		x := args[0].I32()
+		y := args[1].I32()
+
+		return []wasmer.Value{wasmer.NewI32(e + x + y)}, nil
+	},
+)
+```
+
+The biggest improvement is that the environment can be anything! No
+more restriction due to `cgo`. The user is free to store anything
+needed.
+
+### A word about exported functions
+
+The `Exports` API provides the following methods: `GetFunction`,
+`GetMemory`, `GetGlobal` and `GetMemory`, to get an external by its
+name (or simply `Get` to get a generic `Extern`).
+
+`GetFunction` returns an exported function with a native Go API,
+i.e. a function that can be invoked as `addOne(42)`. It's actually an
+alias of `GetRawFunction(name).Native()` (including error handling):
+
+* A _raw function_ is represented by the type `Function`,
+* A _native function_ is represented by the type `NativeFunction`.
+
+The `Function` type provides more information about the function
+itself, whilst `NativeFunction` serves the purpose of being easy to
+use as any Go functions.
+
+```go
+addOne, _ := instance.Exports.GetRawFunction("add_one")
+
+fmt.Println(addOne.Type())
+fmt.Println(addOne.ParameterArity())
+fmt.Println(addOne.ResultArity())
+
+result, _ := addOne.Call(41)
+// or
+addOneNative := addOne.Native()
+result, _ := addOneNative(41)
+```
+
+This example illustrates that even if the new API is more Go
+idiomatic, it's not at the price of losing information or features.
+
+## Compilers and Engines
+
+Let's talk a moment about compilers and engines. Compilers aim at
+compiling the WebAssembly modules into executable codes. Engines drive
+the compilation _and_ the execution of the WebAssembly modules. This
+design provides a unique flexibility which allows Wasmer to be used in
+various contexts.
+
+### Cranelift has new companions: Singlepass and LLVM!
+
+The Wasmer runtime provides 3 compilers to compile the WebAssembly
+modules into executable codes:
+
+* Singlepass: Super fast compilation times, slow execution times. Not
+  prone to JIT-bombs,
+  
+* Cranelift: Fast compilation times, fast execution times,
+
+* LLVM: Slow compilation times, very fast execution times (close to
+  native).
+
+Previously, `wasmer-go` was providing only one compiler:
+Cranelift. However, a non-negligible population of our users run small
+WebAssemby modules. In that specific case, execution time is not an
+issue since it will always be quick, however we must improve the
+compilation time. And that's why we are happy to announce that
+`wasmer-go` now embeds the Singlepass compiler too by default!
+
+Double announce: When execution performance really matters for you,
+you can now use the LLVM compiler too! Even if the default
+`libwasmer`s embedded inside `wasmer-go` do not provide a support for
+LLVM _yet_ (we are working on it), the entire API already support
+it. It is really easy to use your own custom `libwasmer` and build
+against it (see below about the `custom_wasmer_runtime` tag).
+
+Cranelift will continue to be the default compiler. To change that
```

Comment:
Unrelated but the more I think about it the less like calling our compilers `Cranelift` and `LLVM`. Singlepass has a better name even if it's a lie

Reply:
It's not good the marketing yeah.